File: Python/DoublePoint/threeSum/threeSum.py
def threeSum(nums: list[int]) -> list[list[int]]:
    nums.sort()
    res: list[list[int]] = list()

    length = len(nums)

    if length < 3:
        return res

    for index, value in enumerate(nums):
        if index >= length - 2:
            break

        if index > 0 and value == nums[index - 1]:
            continue

        left = index + 1
        right = length - 1

        while left < right:
            sum = value + nums[right] + nums[left]
            if sum == 0:
                res.append([value, nums[left], nums[right]])

                while left < right and nums[left] == nums[left + 1]:
                    left += 1
                while left < right and nums[right] == nums[right - 1]:
                    right -= 1

                left += 1
                right -= 1

            elif sum < 0:
                left += 1
            else:
                right -= 1

    return res

File: java/DoublePoint/threeSum/threeSum.java
import java.util.List;
import java.util.ArrayList;
import java.util.*;

public class threeSum {

    public List<List<Integer>> thrreeSumFunc(int[] nums) {
        int length = nums.length;
        if (length < 3) {
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();

        Arrays.sort(nums);

        for (int i = 0; i < length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = length - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    res.add(new ArrayList<>(List.of(nums[i], nums[left], nums[right])));

                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return res;
    }
}
